cmd/fs-v1-multipart.go

```diff
@@ -18,324 +18,180 @@ package cmd
-		} else {
-			uploadID = upload.UploadID
-			objectName = upload.Object
-			result.Uploads = append(result.Uploads, upload)
+		partPath := pathJoin(uploadIDDir, entry)
+		err = mioutil.AppendFile(file.filePath, partPath)
+		if err != nil {
+			errorIf(err, "Unable to append %s to %s", partPath, file.filePath)
+			return
 		}
-		result.NextKeyMarker = objectName
-		result.NextUploadIDMarker = uploadID
-	}
 
-	if !eof {
-		// Save the go-routine state in the pool so that it can continue from where it left off on
-		// the next request.
-		fs.listPool.Set(listParams{bucket, recursive, result.NextKeyMarker, prefix, heal}, walkResultCh, endWalkCh)
+		file.parts = append(file.parts, PartInfo{PartNumber: nextPartNumber, ETag: etag})
+		nextPartNumber++
 	}
-
-	result.IsTruncated = !eof
-	if !result.IsTruncated {
-		result.NextKeyMarker = ""
-		result.NextUploadIDMarker = ""
-	}
-
-	// Success.
-	return result, nil
 }
 
 // ListMultipartUploads - lists all the uploadIDs for the specified object.
 // We do not support prefix based listing.
-func (fs fsObjects) ListMultipartUploads(bucket, object, keyMarker, uploadIDMarker, delimiter string, maxUploads int) (lmi ListMultipartsInfo, e error) {
+func (fs *fsObjects) ListMultipartUploads(bucket, object, keyMarker, uploadIDMarker, delimiter string, maxUploads int) (result ListMultipartsInfo, e error) {
 	if err := checkListMultipartArgs(bucket, object, keyMarker, uploadIDMarker, delimiter, fs); err != nil {
-		return lmi, err
+		return result, toObjectErr(errors.Trace(err))
 	}
 
 	if _, err := fs.statBucketDir(bucket); err != nil {
-		return lmi, toObjectErr(err, bucket)
+		return result, toObjectErr(errors.Trace(err), bucket)
 	}
 
-	result := ListMultipartsInfo{}
-
 	result.IsTruncated = true
 	result.MaxUploads = maxUploads
 	result.KeyMarker = keyMarker
 	result.Prefix = object
 	result.Delimiter = delimiter
 
-	uploads, _, err := fs.listMultipartUploadIDs(bucket, object, uploadIDMarker, maxUploads)
+	uploadIDs, err := readDir(fs.getMultipartSHADir(bucket, object))
 	if err != nil {
-		return lmi, err
+		if err == errFileNotFound {
+			result.IsTruncated = false
+			return result, nil
+		}
+		return result, toObjectErr(errors.Trace(err))
 	}
 
-	result.NextKeyMarker = object
-	// Loop through all the received uploads fill in the multiparts result.
-	for _, upload := range uploads {
-		uploadID := upload.UploadID
-		result.Uploads = append(result.Uploads, upload)
-		result.NextUploadIDMarker = uploadID
+	// S3 spec says uploaIDs should be sorted based on initiated time.
+	var initiated []time.Time
+	for _, uploadID := range uploadIDs {
+		metaFilePath := pathJoin(fs.getMultipartSHADir(bucket, object), uploadID, fsMetaJSONFile)
+		fi, err := fsStatFile(metaFilePath)
+		if err != nil {
+			return result, toObjectErr(err, bucket, object)
+		}
+		initiated = append(initiated, fi.ModTime())
 	}
+	sort.SliceStable(uploadIDs, func(i int, j int) bool {
```

Comment:
Any particular reason to use sort.SliceStable instead of sort.Slice if yes can you add relevant comment?

Reply:
golang doc says:
```
// Slice sorts the provided slice given the provided less function.                                                                                                                                                
//                                                                                                                                                                                                                 
// The sort is not guaranteed to be stable. For a stable sort, use                                                                                                                                                 
// SliceStable.                                                                                                                                                                                                    
//                                                                                                                                                                                                                 
// The function panics if the provided interface is not a slice.                                                                                                                                                   
func Slice(slice interface{}, less func(i, j int) bool) {                                                                                                                                                          
```

So we should replace all our sort.Slice with the stable version

Reply:
That is incorrect way to look at it. Stable sort is needed when we are sorting based on two or more distinct elements. It is not useful for us and has performance penalty.

When equal elements are indistinguishable, such as with integers, or more generally, any data where the entire element is the key, stability is not an issue. Stability is also not an issue if all keys are different.

Reply:
Nowhere in our code we have requirements for sorting based on two or more distinct elements. For example if you say we sort based on object name and object modified time then stable sort makes sense.

Reply:
I assumed that unstable = unreliable. Will fix this.